libraries/rush-lib/src/logic/pnpm/PnpmfileShim.ts

```diff
@@ -76,27 +86,61 @@ function init(context: IPnpmfileContext | any): IPnpmfileContext {
   return context as IPnpmfileContext;
 }
 
+function parseRange(range: string): TSemver.Range | false {
+  if (!rangeCache || !semver) {
+    return false;
+  }
+
+  if (range.includes(':')) {
+    return false;
+  }
+
+  const entry: TSemver.Range | false | undefined = rangeCache.get(range);
+  if (entry !== undefined) {
+    return entry;
+  }
+
+  try {
+    const parsedRange: TSemver.Range = new semver.Range(range, SEMVER_COMPARE_OPTIONS);
+    rangeCache.set(range, parsedRange);
+    return parsedRange;
+  } catch {
+    rangeCache.set(range, false);
+    return false;
+  }
+}
+
 // Set the preferred versions on the dependency map. If the version on the map is an allowedAlternativeVersion
 // then skip it. Otherwise, check to ensure that the common version is a subset of the specified version. If
 // it is, then replace the specified version with the preferredVersion
 function setPreferredVersions(dependencies: { [dependencyName: string]: string } | undefined): void {
-  for (const [name, version] of Object.entries(dependencies || {})) {
+  if (!dependencies || !semver) {
+    return;
+  }
+
+  for (const [name, version] of Object.entries(dependencies)) {
+    if (version?.includes(':')) {
+      // Skip any specifier that isn't a raw SemVer range.
+      continue;
+    }
+
     const preferredVersion: string | undefined = allPreferredVersions?.get(name);
+    // If preferredVersionRange is valid and the current version is not an allowed alternative, proceed to check subsets
     if (preferredVersion && !allowedAlternativeVersions?.get(name)?.has(version)) {
-      let preferredVersionRange: TSemver.Range | undefined;
+      const preferredVersionRange: TSemver.Range | false = parseRange(preferredVersion);
+      if (!preferredVersionRange) {
+        continue;
+      }
+
       let versionRange: TSemver.Range | undefined;
       try {
-        preferredVersionRange = new semver!.Range(preferredVersion);
-        versionRange = new semver!.Range(version);
+        versionRange = new semver.Range(version, SEMVER_COMPARE_OPTIONS);
```

Comment:
Hmm. You could also cache these versions (in a separate dict) using the function above. Could skip a lot of manual parsing on versions that are common among a lot of packages that need to be mapped to preferred versions.